File: medianOfArrays.cpp
// Two heap solution
// Time Complexity : O(mlogm + nlogn); 
// Space Complexity : O(m+n) 
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// 1. Maintain a max-heap of all smaller elements and a min-heap of all larger elements. The max is allowed one element more than the min-heap
// 2. Add both arrays to the heap such keeping the heaps balanced as per above conditions
// 3. At the end return top of max-heap or mean of top of both heaps depending on their sizes   

class Solution {
public:
    void addToHeap(priority_queue<int>& small, priority_queue<int, vector<int>, greater<int>>& large, vector<int>& nums){
        // small heap can have at the most 1 element more than high 
        for(auto el: nums){
            // cout<<"adding "<<el<<endl; 
            small.push(el);
            if(small.size() == large.size()+2){
                large.push(small.top()); small.pop();
            }
            else if(large.size()!=0 && small.size()>large.size() && small.top()>large.top()){
                int temp1 = small.top(); small.pop();
                int temp2 = large.top(); large.pop();
                large.push(temp1);
                small.push(temp2);
            }
        }
    }
    
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        priority_queue<int> small; // max heap
        priority_queue<int, vector<int>, greater<int>> large; // min heap
        addToHeap(small, large, nums1);
        addToHeap(small, large, nums2);
        if(small.size()>large.size())
            return (double)small.top();
        else
            return (double)(small.top()+large.top())/2;
    }
};

File: MedianofArrays.java
// Time Complexity : O(logm) where m is length of smaller array
// Space Complexity : O(1) No extra space
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : None

class MedainofArrays {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {

        int n1 = nums1.length; int n2 = nums2.length;
        if(n1==0 && n2==0) return 0.0;
        if(n2<n1) return findMedianSortedArrays(nums2,nums1);
        int low = 0;
        int high = n1;
        while(low<=high){
            int px = low + (high-low)/2;
            int py = (n1+n2)/2 - px;
            double l1 = px == 0 ? Integer.MIN_VALUE : nums1[px-1];
            double l2 = py == 0 ? Integer.MIN_VALUE : nums2[py-1];
            double r1 = px == n1 ? Integer.MAX_VALUE : nums1[px];
            double r2 = py == n2 ? Integer.MAX_VALUE : nums2[py];

            if(l1<=r2 && l2<=r1){
                if((n1+n2)%2==0){
                    return ((Math.min(r1,r2)+Math.max(l1,l2))/2);
                }
                else{
                    return Math.min(r1,r2);
                }
            }
            else if(l2>r1){
                low = px+1;
            }
            else{
                high = px-1;
            }
        }
        return 0.0;
    }

    public static void main(String [] args){
        MedainofArrays m = new MedainofArrays();
        System.out.println(m.findMedianSortedArrays(new int[] {1,2}, new int[]{3,4}));
    }
}